presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -716,7 +716,11 @@ public PlanWithProperties visitJoin(JoinNode node, PreferredProperties preferred
             PlanWithProperties right;
 
             JoinNode.DistributionType distributionType = node.getDistributionType().orElseThrow(() -> new IllegalArgumentException("distributionType not yet set"));
-            if (distributionType == JoinNode.DistributionType.PARTITIONED) {
+
+            // exclude replicated join early if probe side is naturally partitioned by left join symbols (e.g: because of aggregation)
+            left = node.getLeft().accept(this, PreferredProperties.any());
```

Comment:
The "addexchanging" of the `left` will be redone few lines down and this might incur additional calls to `getTableLayout`. Would it be possible to avoid this? Looks like the second call (the one requesting particular `PreferredProperties.partitioned(ImmutableSet.copyOf(leftSymbols))` partitioning) will be doing exactly same job most of the time, just adding hashing exchanges on top, no?

(Also, if assignment to `left` is not going to be used except in the `if` right below, avoid the assignment --either use a separate variable or extract to a method

Reply:
I changed the flow a bit. Now we try first preferred partitioning first:
```
            left = node.getLeft().accept(this, PreferredProperties.partitioned(ImmutableSet.copyOf(leftSymbols)));
```
which is then reused in `REPARTITIONED` if branch.

> PreferredProperties.partitioned(ImmutableSet.copyOf(leftSymbols)) partitioning) will be doing exactly same job most of the time, just adding hashing exchanges on top, no?

preferred partitioning does not enforce exchanges.  Enforcing exchanges need to be added explicitly.